File: sdk/streamlib-deno/_manifest.ts
// Copyright (c) 2025 Jonathan Fontanez
// SPDX-License-Identifier: BUSL-1.1

/**
 * Decorator-side reader for `streamlib.yaml` package metadata.
 *
 * Reads only the top-level `package: { org, name, version, ... }` block — the
 * package identity the `@processor` decorator needs to compose a structured
 * `SchemaIdent`.
 *
 * The `processors:` list is NOT read here: the decorator is the truth-source
 * for which processors a package declares (extraction is import — see
 * `extract_processors.ts`). Only the package identity lives in the manifest;
 * the processor set is derived from `@processor` usage in code.
 *
 * The full manifest is parsed by the Rust runtime via `serde_yaml` when the
 * host loads it; this reader only needs the package identity fields.
 *
 * Uses the npm `yaml` parser (declared as a registry dependency) rather
 * than a `jsr:` import: the SDK is published to npm via `deno pack`, and
 * Deno cannot resolve `jsr:` specifiers from inside a published npm
 * package — so the wire-vocabulary deps must be plain npm / `node:`.
 *
 * @module
 */

import { parse as parseYaml } from "yaml";

/** Resolved `package:` block from streamlib.yaml. */
export interface ManifestPackage {
  readonly org: string;
  readonly name: string;
  readonly version: string;
}

/** Raised when a streamlib.yaml cannot be parsed for decorator use. */
export class ManifestParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ManifestParseError";
  }
}

/**
 * Parse the `package:` block from a streamlib.yaml at `path`.
 *
 * Throws `Deno.errors.NotFound` if the file does not exist, and
 * `ManifestParseError` if the manifest is missing `package:`, missing
 * any of `org`/`name`/`version`, or otherwise malformed.
 */
export function readPackageBlock(path: string): ManifestPackage {
  let text: string;
  try {
    text = Deno.readTextFileSync(path);
  } catch (e) {
    if (e instanceof Deno.errors.NotFound) {
      throw e;
    }
    throw new ManifestParseError(
      `failed to read ${path}: ${e instanceof Error ? e.message : String(e)}`,
    );
  }

  let parsed: unknown;
  try {
    parsed = parseYaml(text);
  } catch (e) {
    throw new ManifestParseError(
      `failed to parse ${path}: ${e instanceof Error ? e.message : String(e)}`,
    );
  }

  if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new ManifestParseError(
      `streamlib.yaml at ${path} must be a YAML mapping at the top level`,
    );
  }
  const root = parsed as Record<string, unknown>;

  const pkgRaw = root["package"];
  if (
    pkgRaw === null || pkgRaw === undefined ||
    typeof pkgRaw !== "object" || Array.isArray(pkgRaw)
  ) {
    throw new ManifestParseError(
      `streamlib.yaml at ${path} is missing required \`package:\` block. ` +
        `The decorator requires \`package: { org, name, version }\` to ` +
        `construct a structured SchemaIdent.`,
    );
  }
  const pkg = pkgRaw as Record<string, unknown>;

  const org = pkg["org"];
  const name = pkg["name"];
  const version = pkg["version"];
  const missing: string[] = [];
  if (typeof org !== "string") missing.push("org");
  if (typeof name !== "string") missing.push("name");
  if (typeof version !== "string") missing.push("version");
  if (missing.length > 0) {
    throw new ManifestParseError(
      `streamlib.yaml at ${path} is missing required \`package:\` field(s): ` +
        `${missing.join(", ")}. The decorator requires ` +
        `\`package: { org, name, version }\` to construct a structured SchemaIdent.`,
    );
  }

  return {
    org: org as string,
    name: name as string,
    version: version as string,
  };
}

File: sdk/streamlib-python/python/streamlib/_manifest.py
# Copyright (c) 2025 Jonathan Fontanez
# SPDX-License-Identifier: BUSL-1.1

"""Minimal hand-rolled YAML reader for `streamlib.yaml` package metadata.

Reads only the top-level `package: { org, name, version, ... }` block — the
package identity the `@processor` decorator needs to compose a structured
[`SchemaIdent`][streamlib.schema_ident.SchemaIdent].

The `processors:` list is NOT read here: the decorator is the truth-source
for which processors a package declares (extraction is import — see
[`streamlib.extract_processors`][]). Only the package identity lives in the
manifest; the processor set is derived from `@processor` usage in code.

A full YAML parser would require PyYAML, which would be the SDK's first
runtime dep. The `package:` block shape is constrained enough that a focused
line-oriented reader suffices: 2-space indented mappings, comments, blank
lines, optional single/double quotes around scalars. Anything more complex
(anchors, multi-line strings, flow style outside `{ a: b }`) is out of scope
for this reader.

The full manifest is parsed by the Rust runtime via `serde_yaml` when the
host loads it; this reader only needs the package identity fields.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Optional, Tuple


@dataclass(frozen=True)
class ManifestPackage:
    """Resolved `package:` block from streamlib.yaml."""

    org: str
    name: str
    version: str


class ManifestParseError(ValueError):
    """Raised when a streamlib.yaml cannot be parsed for decorator use."""


def read_package_block(path: Path) -> ManifestPackage:
    """Parse the `package:` block from a streamlib.yaml.

    Raises:
        FileNotFoundError: if `path` does not exist.
        ManifestParseError: if the manifest is missing `package:`, or missing
            any of `org`/`name`/`version`, or malformed.
    """
    if not path.exists():
        raise FileNotFoundError(str(path))

    text = path.read_text(encoding="utf-8")
    org, name, version = _scan_package(text)

    missing = [field for field, val in (("org", org), ("name", name), ("version", version)) if val is None]
    if missing:
        raise ManifestParseError(
            f"streamlib.yaml at {path} is missing required `package:` field(s): "
            f"{', '.join(missing)}. The decorator requires `package: {{ org, name, version }}` "
            f"to construct a structured SchemaIdent."
        )

    return ManifestPackage(org=org, name=name, version=version)


def _scan_package(text: str) -> Tuple[Optional[str], Optional[str], Optional[str]]:
    org: Optional[str] = None
    name: Optional[str] = None
    version: Optional[str] = None

    in_package = False

    for raw in text.splitlines():
        line = _strip_comment(raw).rstrip()
        if not line.strip():
            continue

        indent = len(line) - len(line.lstrip())
        content = line.lstrip()

        if indent == 0:
            in_package = content.rstrip(":") == "package" and content.endswith(":")
            continue

        if in_package and indent == 2 and ":" in content:
            key, _, val = content.partition(":")
            key = key.strip()
            val = _strip_quotes(val.strip())
            if key == "org":
                org = val
            elif key == "name":
                name = val
            elif key == "version":
                version = val

    return org, name, version


def _strip_quotes(s: str) -> str:
    """Strip a single surrounding pair of double or single quotes."""
    if len(s) >= 2 and ((s[0] == s[-1] == '"') or (s[0] == s[-1] == "'")):
        return s[1:-1]
    return s


def _strip_comment(s: str) -> str:
    """Strip a trailing `# ...` comment when `#` is whitespace-preceded.

    Plays it safe around `#` characters embedded in string values: only
    strips when the `#` follows whitespace and is outside of a quoted
    region.
    """
    in_quote: Optional[str] = None
    for i, c in enumerate(s):
        if c in ('"', "'"):
            if in_quote is None:
                in_quote = c
            elif in_quote == c:
                in_quote = None
        elif c == "#" and in_quote is None:
            if i == 0 or s[i - 1].isspace():
                return s[:i].rstrip()
    return s
